openapi-codegen/core/src/main/scala/sttp/tapir/codegen/ValidationGenerator.scala

```diff
@@ -0,0 +1,349 @@
+package sttp.tapir.codegen
+
+import io.circe.JsonNumber
+import sttp.tapir.codegen.openapi.models.OpenapiModels.OpenapiDocument
+import sttp.tapir.codegen.openapi.models.OpenapiSchemaType
+import sttp.tapir.codegen.openapi.models.OpenapiSchemaType.{
+  OpenapiSchemaArray,
+  OpenapiSchemaMap,
+  OpenapiSchemaNumericType,
+  OpenapiSchemaObject,
+  OpenapiSchemaRef,
+  OpenapiSchemaString
+}
+import sttp.tapir.codegen.util.{DocUtils, JavaEscape}
+
+import scala.annotation.tailrec
+
+case class ValidationDefn(name: String, tpe: String, construct: Set[String] => Option[String], refOnly: Boolean = false)
+case class ValidationDefns(defns: Map[String, ValidationDefn]) {
+  def render: String = defns.values
+    .flatMap { d => d.construct(defns.keySet).map(impl => s"lazy val ${d.name}Validator: Validator[${d.tpe}] = $impl") }
+    .mkString("\n")
+}
+object ValidationDefns {
+  val empty: ValidationDefns = ValidationDefns(Map.empty)
+  @tailrec
+  private[ValidationDefns] def filtered(defns: ValidationDefns): ValidationDefns = {
+    val next = ValidationDefns(defns.defns.filter(_._2.construct(defns.defns.keySet).isDefined))
+    if (next.defns.keySet == defns.defns.keySet) defns
+    else filtered(next)
+  }
+}
+
+object ValidationGenerator {
+  private def render(n: JsonNumber): String = n.toLong.map(_.toString).getOrElse(n.toDouble.toString)
+  private def allowNull(tpe: String, required: Boolean)(s: String): String = if (required) s
+  else
+    s"""Validator.custom[Option[$tpe]](ot => ot.map($s(_)).map{
+       |  case Nil => ValidationResult.Valid
+       |  case l   => ValidationResult.Invalid(l.flatMap(_.customMessage).toList)
+       |}.getOrElse(ValidationResult.Valid))""".stripMargin
+  private def opt(t: String, nullable: Boolean) = if (nullable) s"Option[$t]" else t
+  private def singleton(name: String, tpe: String, validationDefn: String): Seq[ValidationDefn] =
+    Seq(ValidationDefn(name, tpe, _ => Some(validationDefn)))
+
+  def mkValidators(doc: OpenapiDocument): ValidationDefns = {
+    val allSchemas = doc.components.map(_.schemas).getOrElse(Map.empty)
+
+    // All schemas that have explicit validation _or_ refer to a ref, which _may_ have.
+    // We need to filter this, because we may have trivial
+    val unfiltered = {
```

Comment:
I hate this bit. We traverse the doc tree way too many times as it stands in this generator. What we really need is a tree of all schemas that reference a given one, and vice-verse, that we can then reuse between various places that rebuild a small part of the world.